### PR TITLE
dart-sdk 2.17.3 (new formula)

### PR DESCRIPTION
**Pull request by SMillerDev**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----
I'm not happy with the usage of fetch but for now this just follows upstream instructions.

### CODE REVIEW

**Comment by SMillerDev** [comment]
That's nice, the template tests fail on ARM...

**Comment by bayandin** [comment]
> I'm not happy with the usage of fetch but for now this just follows upstream instructions.

We could probably do something like we have for `v8` formula — to list a bunch of resources and install them during the build.

**Comment by SMillerDev** [comment]
Yeah, I was thinking of making a `bump-deps-resources` command for that already. Would be a nice improvement.

File: Formula/dart-sdk.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  def install
+    resource("depot-tools").stage(buildpath/"depot-tools")
+
+    ENV.append_path "PATH", "#{buildpath}/depot-tools"
+    system "fetch", "--no-history", "dart"
+    chdir "sdk" do
+      arch = Hardware::CPU.arm? ? "arm64" : "x64"
+      system "./tools/build.py", "--no-goma", "--mode=release", "--arch=#{arch}", "create_sdk"
+      out = OS.linux? ? "out" : "xcodebuild"
+      prefix.install Dir["#{out}/Release#{arch}/dart-sdk/*"]
```
Comment (SMillerDev) [inline Formula/dart-sdk.rb:28]:
```suggestion
      prefix.install Dir["#{out}/Release#{arch.capitalize}/dart-sdk/*"]
```

**Comment by SMillerDev** [comment]
Okay, this now seems ready for review. I'm planning to replace fetch usage later, when I have a command to resolve DEPS files

**Review by dawidd6 — APPROVED** [review]

**Review by bayandin — APPROVED** [review]

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by mit-mit** [comment]
Hey @SMillerDev @bayandin. Michael and Alex (cc @athomas) from the Dart team here.

We've had this Dart homebrew formula for a while:
https://github.com/dart-lang/homebrew-dart/blob/master/Formula/dart.rb

We really like the idea of having Dart support in homebrew core though, so we can avoid the extra `brew tap` step.

The one we wrote installs the Dart SDK by pulling down our official, prebuilt, signed SDKs. This new one builds from source. I suspect that's due to the general preference to not have binary formulae? 

Building Dart from source was designed for those contributing to the SDK, so would an alternative be to use a cask that pulls the prebuilt binaries?

**Comment by bayandin** [comment]
Hey @mit-mit,

> I suspect that's due to the general preference to not have binary formulae?

Yep, it's one of the main homebrew-core requirements.

> Building Dart from source was designed for those contributing to the SDK, so would an alternative be to use a cask that pulls the prebuilt binaries?

Usually, we recommend using a custom tap for such cases, so [`dart-lang/dart/dart`](https://github.com/dart-lang/homebrew-dart/blob/master/Formula/dart.rb) is a perfect example.

**Comment by carlocab** [comment]
Hi, @mit-mit, thanks for dropping by.

As @bayandin points out, this is a much better fit for an external tap. (It's possible shipping your existing formulae as casks might be a better UX for your users; but I haven't thought about it too carefully.)

Per our documentation on [acceptable casks](https://docs.brew.sh/Acceptable-Casks#rejected-casks), one common reason for rejecting a cask is

> The app is both open-source and CLI-only (i.e. it only uses the binary artifact). In that case, and [in the spirit of deduplication](https://github.com/Homebrew/homebrew-cask/issues/15603), submit it first to [Homebrew/core](https://github.com/Homebrew/homebrew-core) as a formula that builds from source. If it is rejected, you may then try again as a cask (link us to the issue so we can see the discussion and reasoning for rejection).

There are, of course, exceptions to the above. However, the very nature of exceptions means that we would be very reluctant to add another one.

File: Formula/dart-sdk.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  depends_on "ninja" => :build
+  depends_on "rust" => :build
+
+  uses_from_macos "curl" => :build
+  uses_from_macos "python" => :build
+
+  resource "depot-tools" do
+    url "https://chromium.googlesource.com/chromium/tools/depot_tools.git",
+        revision: "138bff2823590b3f3db440425bf712392defb7de"
+  end
```
Comment (chenrui333) [inline Formula/dart-sdk.rb:17]: is this always using the latest revision of the `depot_tools`?

Comment (iMichka) [reply Formula/dart-sdk.rb:17]: I guess yes. They don't seem to have tag, so better than nothing.
I am wondering if we really need to ship that dependency ...

Comment (SMillerDev) [reply Formula/dart-sdk.rb:17]: AFAIK we don't ship the tools themselves. It's just needed to build.

**Pull request closed without merging**